hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java

```diff
@@ -1290,7 +1285,7 @@ private void addParentChildEntityNameByPropertyAndValue(AbstractEntityInsertActi
 				final String entityName = entityType.getName();
 				final String rootEntityName = action.getSession().getFactory().getMetamodel().entityPersister( entityName ).getRootEntityName();
 
-				if ( entityType.isOneToOne() && OneToOneType.class.cast( entityType ).getForeignKeyDirection() == ForeignKeyDirection.TO_PARENT ) {
+				if ( entityType.isOneToOne() && entityType.getForeignKeyDirection() == ForeignKeyDirection.TO_PARENT ) {
```

Comment:
We don't need to down cast to invoke the method for Java's polymorphism is builtin feature. Maybe the casting enjoys better performance?

Reply:
yes, we have occasionally needed an explicit cast introduced by the performance team. I don't remember if this is the case, but it would be safer to not clean these up blindly: an excplicit cast has the ability to simplify a megamorphic call and we have a couple of extreme cases.

Reply:
reverted it back.